x-pack/plugins/security_solution/public/detection_engine/rule_creation/components/required_fields/required_fields.tsx

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import React from 'react';
+import { EuiButtonEmpty, EuiCallOut, EuiFormRow, EuiSpacer, EuiText } from '@elastic/eui';
+import type { DataViewFieldBase } from '@kbn/es-query';
+import type { RequiredFieldInput } from '../../../../../common/api/detection_engine';
+import { UseArray, useFormData } from '../../../../shared_imports';
+import type { ArrayItem } from '../../../../shared_imports';
+import { RequiredFieldRow } from './required_fields_row';
+import * as ruleDetailsI18n from '../../../rule_management/components/rule_details/translations';
+import * as i18n from './translations';
+
+interface RequiredFieldsProps {
+  path: string;
+  indexPatternFields?: DataViewFieldBase[];
+  isIndexPatternLoading?: boolean;
+}
+
+export const RequiredFields = ({
+  path,
+  indexPatternFields = [],
+  isIndexPatternLoading = false,
+}: RequiredFieldsProps) => (
+  <UseArray path={path} initialNumberOfItems={0}>
+    {({ items, addItem, removeItem }) => (
+      <RequiredFieldsList
+        items={items}
+        addItem={addItem}
+        removeItem={removeItem}
+        indexPatternFields={indexPatternFields}
+        isIndexPatternLoading={isIndexPatternLoading}
+        path={path}
+      />
+    )}
+  </UseArray>
+);
+
+interface RequiredFieldsListProps {
+  items: ArrayItem[];
+  addItem: () => void;
+  removeItem: (id: number) => void;
+  indexPatternFields: DataViewFieldBase[];
+  isIndexPatternLoading: boolean;
+  path: string;
+}
+
+const RequiredFieldsList = ({
+  items,
+  addItem,
+  removeItem,
+  indexPatternFields,
+  isIndexPatternLoading,
+  path,
+}: RequiredFieldsListProps) => {
+  const useFormDataResult = useFormData();
+  const [formData] = useFormDataResult;
+  const fieldValue: RequiredFieldInput[] = formData[path] ?? [];
+
+  const selectedFieldNames = fieldValue.map(({ name }) => name);
+
+  const fieldsWithTypes = indexPatternFields.filter(
+    (indexPatternField) => indexPatternField.esTypes && indexPatternField.esTypes.length > 0
+  );
+
+  const allFieldNames = fieldsWithTypes.map(({ name }) => name);
+  const availableFieldNames = allFieldNames.filter((name) => !selectedFieldNames.includes(name));
```

Comment:
I have investigated this and have concluded that I can't `useMemo` `availableFieldNames` and other related variables because they depend on `fieldValue` and `fieldValue` has a different reference on every render. That's because `useFormData` recreates array field values on every form data change.

But I fixed the issue in a different way by passing the "watch" arg to `useMemo` which makes it re-render only when either "index patterns" or "required fields" themselves change. And if these change, we'll have to compute these values anyways. Feel free to ask questions if it doesn't make sense.